NDTensors/src/lib/UnallocatedArrays/test/runtests.jl

```diff
@@ -0,0 +1,207 @@
+    # UnallocatedFill
+    f = Fill{elt}(3.0, (2, 3, 4))
+    F = UnallocatedFill(f, Array{elt,ndims(f)})
+    @test F isa AbstractFill
+    @test size(F) == (2, 3, 4)
+    @test length(F) == 24
+    @test sum(F) ≈ 3 * 24
+    @test norm(F) ≈ sqrt(3^2 * 24)
+    @test F[2, 3, 1] == 3.0
+    @test allocate(F) isa Array{elt,3}
+    Fp = allocate(F)
+    @test norm(Fp) ≈ norm(F)
+    Fs = similar(F)
+    @test Fs isa alloctype(F)
+    @test Fs[1, 1, 1] != 3.0
+
+    Fp = set_alloctype(f, dev(Array{elt,ndims(f)}))
+    @test allocate(Fp) isa dev(Array{elt,ndims(f)})
+    @test Fp == F
+    Fc = copy(F)
+    @test Fc == F
+    Fc = allocate(complex(F))
+    @test eltype(Fc) == complex(eltype(F))
+    @test typeof(Fc) == alloctype(complex(F))
+    Fc[2, 3, 4] = 4.0 + 3.0im
+    @test Fc[2, 3, 4] == 4.0 + 3.0im
+    ## TODO this isn't working
+    #Fc[2,3,4] = Base.setindex(F, 4.0+3.0im, 2,3,4)
+  end
+
+  @testset "Multiplication" begin
+    z = Zeros{elt}((2, 3))
+    Z = UnallocatedZeros(z, dev(Matrix{eltype(z)}))
+
+    R = Z * Z'
+    @test R isa UnallocatedZeros
+    @test alloctype(R) == alloctype(Z)
+    @test size(R) == (2, 2)
+    M = rand(elt, (3, 4))
+    R = Z * M
+    @test R isa UnallocatedZeros
+    @test alloctype(R) == alloctype(Z)
+    @test size(R) == (2, 4)
+    R = M' * Z'
+    @test R isa UnallocatedZeros
+    @test alloctype(R) == alloctype(Z)
+    @test size(R) == (4, 2)
+    R = transpose(M) * transpose(Z)
+    @test R isa UnallocatedZeros
+    @test alloctype(R) == alloctype(Z)
+    @test size(R) == (4, 2)
+
+    ###################################
+    ## UnallocatedFill
+    f = Fill{elt}(3.0, (2, 12))
+    F = UnallocatedFill(f, dev(Matrix{elt}))
+    p = Fill{elt}(4.0, (12, 5))
+    P = UnallocatedFill(p, dev(Array{elt,ndims(p)}))
+    R = F * P
+    @test F isa UnallocatedFill
+    @test R[1, 1] == 144
+    @test alloctype(R) == alloctype(F)
+    @test size(R) == (2, 5)
+
+    R = F * F'
+    @test R isa UnallocatedFill
+    @test R[1, 2] == elt(108)
+    @test alloctype(R) == alloctype(F)
+    @test size(R) == (2, 2)
+
+    R = transpose(F) * F
+    @test R isa UnallocatedFill
+    @test R[12, 3] == elt(18)
+    @test alloctype(R) == alloctype(F)
+    @test size(R) == (12, 12)
+
+    R = transpose(Z) * F
+    @test R isa UnallocatedZeros
+    @test alloctype(R) == alloctype(Z)
+    @test size(R) == (3, 12)
+  end
+
+  @testset "Broadcast" begin
+    z = Zeros{elt}((2, 3))
+    Z = UnallocatedZeros(z, dev(Matrix{elt}))
+    R = elt(2.0) .* Z
+    @test R isa UnallocatedZeros
+    @test alloctype(R) == alloctype(Z)
+    R = Z .* elt(2.0)
+    @test R isa UnallocatedZeros
+    @test alloctype(R) == alloctype(Z)
+
+    R = Z .* Z
+    @test R isa UnallocatedZeros
+    @test alloctype(R) == alloctype(Z)
+
+    Z = UnallocatedZeros(Zeros{elt}((2, 3)), dev(Matrix{elt}))
+    R = Z + Z
+    @test R isa UnallocatedZeros
+    @test alloctype(R) == alloctype(Z)
+
+    R = Z .+ elt(2.0)
```

Comment:
```suggestion
    R = Z .+ elt(2)
```

Reply:
This makes sense, especially because we are checking for Float32, so I could be accidentally promoting to `Float64`. I converted all of the instances of Float64 numbers (x.0) to ints.